x/distribution/keeper/allocation.go

```diff
@@ -51,35 +51,63 @@ func (k Keeper) AllocateTokens(ctx context.Context, totalPreviousPower int64, bo
 	voteMultiplier := math.LegacyOneDec().Sub(communityTax)
 	feeMultiplier := feesCollected.MulDecTruncate(voteMultiplier)
 
-	// allocate tokens proportionally to voting power
-	//
-	// TODO: Consider parallelizing later
-	//
-	// Ref: https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r246276376
-	for _, vote := range bondedVotes {
+	var wg sync.WaitGroup
+	// channel for errors
+	errCh := make(chan error, len(bondedVotes))
+	// channel for remaining rewards
+	remainingCh := make(chan sdk.DecCoins, len(bondedVotes))
 
-		validator, err := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
-		if err != nil {
-			return err
-		}
+	for _, vote := range bondedVotes {
+		wg.Add(1)
+		go func(vote comet.VoteInfo) {
+			defer wg.Done()
+
+			validator, err := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
+			if err != nil {
+				errCh <- err
+				return
+			}
+
+			powerFraction := math.LegacyNewDec(vote.Validator.Power).QuoTruncate(math.LegacyNewDec(totalPreviousPower))
+			reward := feeMultiplier.MulDecTruncate(powerFraction)
+
+			if err = k.AllocateTokensToValidator(ctx, validator, reward); err != nil {
```

Comment:
As it is getting different validators, I thought there would not be concurrency issues. Additionally, events are indeed not in consensus, so I thought it would be fine.